src/test/kotlin/org/rust/ide/annotator/fixes/AddTurbofishFixTest.kt

```diff
@@ -48,14 +47,54 @@ class AddTurbofishFixTest : RsAnnotatorTestBase() {
         }
         """)
 
-    fun `test add turbofish should not be available if the instance is corrected yet`() = checkNoIntention(intention,
+    fun `test should recognize also generics arguments`() {
+        for (gen in listOf("<Option<String>>",
+            "<Option<Option<String>>>",
+            "<Option<Option<Option<String>>>>",
+            "<Option<Option<Option<Option<String>>>>>")) {
+            checkQuickFix(
+                """
+                    fn foo() {
+                        let _ = something$gen(/*caret*/"42");
+                    }
+                """, """
+                    fn foo() {
+                        let _ = something::$gen(/*caret*/"42");
+                    }
+                """)
+        }
+    }
+
+    fun `test should guess the correct boundary`() {
+        val cases = listOf(
+            Pair("", " >> 3"), Pair("", " > 3 > 5"), Pair("", " == call_something()"),
+            Pair("3 < ", ""), Pair("3 == ", ""), Pair("bye() == 3 < ", " > more() >> 2")
+        )
+
+        for ((before, after) in cases) {
```

Comment:
I think this loop is a bit too unreadable. What do you think about creating a function that checks two expressions, and then just write a large number of cases explicitelly?

```Rust
fun doCheck(expr1: String, expr2: String) { ... }

fun `test should guess correct boundary` {
  doCheck("something<Option<i32>>(/*caret*/"42")>> 3", "something::<Option<i32>>(/*caret*/"42")>> 3")
  ...
}
```

It's a bit more repetition that way, but it is super obvious what exactly is expression before and after.